### ISSUE TITLE
Imgui crashes "Assertion failed", delta-t < 0.0

### ISSUE THREAD

**Issue by Atomschlag** [pioneerspacesim/pioneer#3840]
With an config file present the game crashes on startup
 "Assertion failed! pioneer.exe file: imgui.cpp, line 2085
Expression g.IO.DeltaTime >= 0.0f"

how to reproduce: change resolution or graphic settings and restart game

game starts fine in 800x600 widowed with no ini file present

fullscreen mode seems to trigger the issue

chrash apperars mostly when trying to switch to fullscreen
win 10 home
8gb ram

**Comment by ecraven** [comment]
I'm not sure what the problem is here, it *seems* that `io.DeltaTime` is < 0.0, which I do not understand, as it is set to the difference of SDL_GetTicks() / 1000.0 and the previous frame each frame.
Could you elaborate on what you were doing when it happened, and how long you were playing?

**Comment by ecraven** [comment]
https://github.com/ecraven/pioneer/tree/fix-imgui-zero-delta-t contains a single commit over master that prints more detailed stats. Could anyone who is experiencing these problems try with that branch?

**Comment by impaktor** [comment]
For reference:
```
<JohnJ> SDL_GetTicks can go backwards, apparently                       [01:10]
<JohnJ> caused by Linux kernel glitch                                   [01:15]
<JohnJ> (and shitty x86 timer hardware)
```

**Comment by ecraven** [comment]
Unfortunately, both atomschlag and LjL (who reported the same problem) are running on windows, not Linux.

**Comment by jaj22** [comment]
Yeah, that was out of date. Backwards time was a common problem on Linux around 14 years ago. 

Apparently it's a problem on whatever atomschlag and LjL are using. LjL has a Phenom 720 on Windows 10, and those early-ish phenoms are relatively likely to have RDTSC sync bugs. The exact cause doesn't really matter though: If the timer isn't guaranteed to be monotonic then deltas derived from it should be capped to zero. There's no harm in doing so for our purposes.

A zero delta cap should also be applied to the main loop (in Pi.cpp). Amusingly, negative deltas don't currently appear to break anything outside imgui, but that's an unnecessary risk.

**Comment by ecraven** [comment]
See https://github.com/pioneerspacesim/pioneer/pull/3854

**Comment by jaj22** [comment]
Pioneer's internal delta-t also uses successive SDL_GetTicks() calls. It's simply luck if nothing else falls over, as the assumption when writing the code was delta-t >= 0.

I assumed the eventual intention was to merge an imgui backend into the current rendering functionality rather than having a lump of GL code outside, but that's a separate issue.

**Comment by fluffyfreak** [comment]
> I assumed the eventual intention was to merge an imgui backend into the current rendering functionality rather than having a lump of GL code outside, but that's a separate issue.

Yes that's my intention, on hold until we need it basically :)

**Comment by laarmen** [comment]
I believe this has been fixed by #3854 .
